### PR TITLE
virttest.qemu_monitor: add cmd "transaction".

### PR DESCRIPTION
**Pull request by CongLi**
Signed-off-by: Cong Li <coli@redhat.com>

id: 1321839

### CODE REVIEW

**Comment by CongLi** [comment]
17:05:51 DEBUG| (monitor qmpmonitor1) Sending command 'transaction'
17:05:51 DEBUG| Send command: {'execute': 'transaction', 'arguments': {'actions': [{'data': {'device': u'drive_image1', 'snapshot-file': '/usr/share/avocado/data/avocado-vt/images/None', 'mode': 'absolute-paths', 'format': 'qcow2'}, 'type': 'blockdev-snapshot-sync'}]}, 'id': 'qDkFzh5Y'}
17:05:51 INFO | [qemu output] Formatting '/usr/share/avocado/data/avocado-vt/images/None', fmt=qcow2 size=21474836480 backing_file=/usr/share/avocado/data/avocado-vt/images/RHEL-Server-7.3-64-virtio.qcow2 backing_fmt=qcow2 encryption=off cluster_size=65536 lazy_refcounts=off refcount_bits=16

**Comment by CongLi** [comment]
17:01:07 DEBUG| Send command: {'execute': 'transaction', 'arguments': {'actions': [{'data': {'device': u'drive_image1', 'snapshot-file': '/usr/share/avocado/data/avocado-vt/images/image1-snap', 'mode': 'absolute-paths', 'format': 'qcow2'}, 'type': 'blockdev-snapshot-sync'}, {'data': {'device': u'drive_sn1', 'snapshot-file': '/usr/share/avocado/data/avocado-vt/images/sn1-snap', 'mode': 'absolute-paths', 'format': 'qcow2'}, 'type': 'blockdev-snapshot-sync'}, {'data': {'device': u'drive_sn2', 'snapshot-file': '/usr/share/avocado/data/avocado-vt/images/sn2-snap', 'mode': 'absolute-paths', 'format': 'qcow2'}, 'type': 'blockdev-snapshot-sync'}]}, 'id': 'jXuHTrBa'}
17:01:07 INFO | [qemu output] Formatting '/usr/share/avocado/data/avocado-vt/images/image1-snap', fmt=qcow2 size=21474836480 backing_file=/usr/share/avocado/data/avocado-vt/images/RHEL-Server-7.3-64-virtio.qcow2 backing_fmt=qcow2 encryption=off cluster_size=65536 lazy_refcounts=off refcount_bits=16
17:01:07 INFO | [qemu output] Formatting '/usr/share/avocado/data/avocado-vt/images/sn1-snap', fmt=qcow2 size=21474836480 backing_file=/usr/share/avocado/data/avocado-vt/images/sn1.qcow2 backing_fmt=qcow2 encryption=off cluster_size=65536 lazy_refcounts=off refcount_bits=16
17:01:07 INFO | [qemu output] Formatting '/usr/share/avocado/data/avocado-vt/images/sn2-snap', fmt=qcow2 size=21474836480 backing_file=/usr/share/avocado/data/avocado-vt/images/sn2.qcow2 backing_fmt=qcow2 encryption=off cluster_size=65536 lazy_refcounts=off refcount_bits=16

File: virttest/qemu_monitor.py [hunk 1]
```diff
@@ -2210,3 +2210,15 @@ def system_powerdown(self):
...
         logging.info("%s QMP event received" % qmp_event)
+
+    def transaction(self, args):
+        """
+        Atomically operate on a group of one or more block devices.
+
+        :param args: information needed for the device snapshot.
+
+        :return: nothing on success
+                 If @device is not a valid block device, DeviceNotFound
```
Comment (PandaWei) [inline virttest/qemu_monitor.py]: LGTM, what's the meaning of @device. Forgive my ignorance

Comment (CongLi) [inline virttest/qemu_monitor.py]: This is picked up from source code of qemu-kvm.
device should be defined in "BlockdevAction".
I can remove that line if it's confused.

{ 'command': 'transaction',
  'data': { 'actions': [ 'BlockdevAction' ] } }

#  List of:
#  @BlockdevAction: information needed for the device snapshot
#
# Returns: nothing on success
#          If @device is not a valid block device, DeviceNotFound

File: virttest/qemu_monitor.py [hunk 2]
```diff
@@ -2210,3 +2210,15 @@ def system_powerdown(self):
         if not utils_misc.wait_for(lambda: self.get_event(qmp_event), 120):
             raise QMPEventError(cmd, qmp_event, self.name)
         logging.info("%s QMP event received" % qmp_event)
+
+    def transaction(self, args):
```
Comment (xutian) [inline virttest/qemu_monitor.py]: args of transaction should be a list, a job list.

File: virttest/qemu_monitor.py [hunk 3]
```diff
@@ -2210,3 +2210,18 @@ def system_powerdown(self):
         if not utils_misc.wait_for(lambda: self.get_event(qmp_event), 120):
             raise QMPEventError(cmd, qmp_event, self.name)
         logging.info("%s QMP event received" % qmp_event)
+
+    def transaction(self, job_list):
+        """
+        Atomically operate on a group of one or more block devices.
+
+        :param job_list: List of information needed for the device snapshot.
```
Comment (xutian) [inline virttest/qemu_monitor.py]: block_jobs command list. right ?

**Comment by suqinhuang** [comment]
LGTM
Acked-by: Suqin Huang

**Pull request merged**